x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,275 @@
+	suite.chainB = &osmosisibctesting.TestChain{
+		TestChain: suite.coordinator.GetChain(ibctesting.GetChainID(2)),
+	}
+	suite.path = NewTransferPath(suite.chainA, suite.chainB)
+	suite.coordinator.Setup(suite.path)
+}
+
+// Helpers
+func (suite *MiddlewareTestSuite) NewValidMessage(forward bool, amount sdk.Int) sdk.Msg {
+	var coins sdk.Coin
+	var port, channel, accountFrom, accountTo string
+
+	if forward {
+		coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)
+		port = suite.path.EndpointA.ChannelConfig.PortID
+		channel = suite.path.EndpointA.ChannelID
+		accountFrom = suite.chainA.SenderAccount.GetAddress().String()
+		accountTo = suite.chainB.SenderAccount.GetAddress().String()
+	} else {
+		coins = transfertypes.GetTransferCoin(
+			suite.path.EndpointB.ChannelConfig.PortID,
+			suite.path.EndpointB.ChannelID,
+			sdk.DefaultBondDenom,
+			sdk.NewInt(1),
+		)
+		coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)
+		port = suite.path.EndpointB.ChannelConfig.PortID
+		channel = suite.path.EndpointB.ChannelID
+		accountFrom = suite.chainB.SenderAccount.GetAddress().String()
+		accountTo = suite.chainA.SenderAccount.GetAddress().String()
+	}
+
+	timeoutHeight := clienttypes.NewHeight(0, 100)
+	return transfertypes.NewMsgTransfer(
+		port,
+		channel,
+		coins,
+		accountFrom,
+		accountTo,
+		timeoutHeight,
+		0,
+	)
+}
+
+func (suite *MiddlewareTestSuite) ExecuteReceive(msg sdk.Msg) (string, error) {
+	res, err := suite.chainB.SendMsgsNoCheck(msg)
+	suite.Require().NoError(err)
+
+	packet, err := ibctesting.ParsePacketFromEvents(res.GetEvents())
+	suite.Require().NoError(err)
+
+	err = suite.path.EndpointA.UpdateClient()
+	suite.Require().NoError(err)
+
+	res, err = suite.path.EndpointA.RecvPacketWithResult(packet)
+	suite.Require().NoError(err)
+
+	ack, err := ibctesting.ParseAckFromEvents(res.GetEvents())
+	return string(ack), err
+}
+
+func (suite *MiddlewareTestSuite) AssertReceiveSuccess(success bool, msg sdk.Msg) (string, error) {
+	ack, err := suite.ExecuteReceive(msg)
+	if success {
+		suite.Require().NoError(err)
+		suite.Require().NotContains(string(ack), "error",
+			"acknoledgment is an error")
+	} else {
+		suite.Require().Contains(string(ack), "error",
+			"acknoledgment is not an error")
+		suite.Require().Contains(string(ack), types.RateLimitExceededMsg,
+			"acknoledgment error is not of the right type")
+	}
+	return ack, err
+}
+
+func (suite *MiddlewareTestSuite) AssertSendSuccess(success bool, msg sdk.Msg) (*sdk.Result, error) {
+	r, err := suite.chainA.SendMsgsNoCheck(msg)
+	if success {
+		suite.Require().NoError(err, "IBC send failed. Expected success. %s", err)
+	} else {
+		suite.Require().Error(err, "IBC send succeeded. Expected failure")
+		suite.ErrorContains(err, types.RateLimitExceededMsg, "Bad error type")
+	}
+	return r, err
+}
+
+func (suite *MiddlewareTestSuite) BuildChannelQuota(name string, duration, send_precentage, recv_percentage uint32) string {
+	return fmt.Sprintf(`
```

Comment:
yeah, we could create a struct and marshal it to json. Mostly left it as a string here for simplicity since (1) this is only used within a test, and (2) when we call the contract from the cli we use the text json format, so it's easy to copy/paste them to test

Reply:
I generally found the strings to be hard to manage in tests in the SDK, especially when the interfaces keep getting updated and when there are many of them. However, not feeling strongly about this, and the 2) point makes sense so let's try the strings

Reply:
tried this but needed a lot of boilerplate for type conversions, so leaving as strings for now